cdap-api/src/main/java/co/cask/cdap/api/plugin/Plugin.java

```diff
@@ -18,24 +18,44 @@
 
 import co.cask.cdap.api.artifact.ArtifactId;
 
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
 import java.util.Objects;
 
 /**
  * A container class for holding plugin information.
  */
 public final class Plugin {
+  private final List<ArtifactId> parents;
   private final ArtifactId artifactId;
   private final PluginClass pluginClass;
   private final PluginProperties properties;
 
   public Plugin(ArtifactId artifactId, PluginClass pluginClass, PluginProperties properties) {
+    this(new ArrayList<ArtifactId>(), artifactId, pluginClass, properties);
+  }
+
+  public Plugin(List<ArtifactId> parents, ArtifactId artifactId,
+                PluginClass pluginClass, PluginProperties properties) {
+    this.parents = Collections.unmodifiableList(new ArrayList<>(parents));
     this.artifactId = artifactId;
     this.pluginClass = pluginClass;
     this.properties = properties;
   }
 
   /**
-   * @return artifact id
+   * @return the artifact parent chain. Every artifact in the list is the parent of the artifact behind it.
+   *         This list does not contain the plugin's artifact, which means it will be an empty list if the plugin's
+   *         parent is a program and not another plugin.
+   */
+  public List<ArtifactId> getParents() {
+    // null check for backwards compatibility, since this field was added in 4.3.0
+    return parents == null ? new ArrayList<ArtifactId>() : parents;
```

Comment:
Use `Collection.emptyList()` instead so that the list returned is always unmodifiable.